File: push/src/main/java/com/xiaomi/slim/Blob.java
package com.xiaomi.slim;

import android.text.TextUtils;
import com.xiaomi.channel.commonutils.logger.MyLog;
import com.xiaomi.mipush.sdk.Constants;
import com.xiaomi.push.protobuf.ChannelMessage;
import com.xiaomi.push.service.RC4Cryption;
import com.xiaomi.smack.packet.Packet;
import com.xiaomi.smack.util.StringUtils;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.nio.ByteBuffer;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/slim/Blob.class */
public class Blob {
    static final int CHCKSUM_SIZE = 4;
    static final int CIPHER_AES = 2;
    static final int CIPHER_NONE = 0;
    static final int CIPHER_RC4 = 1;
    public static final String CLIENT_PING_ID = "0";
    public static final String CMD_BIND = "BIND";
    public static final String CMD_CLOSE = "CLOSE";
    public static final String CMD_CONN = "CONN";
    public static final String CMD_KICK = "KICK";
    public static final String CMD_NOTIFY = "NOTIFY";
    public static final String CMD_PING = "PING";
    public static final String CMD_SECMSG = "SECMSG";
    public static final String CMD_SYNC = "SYNC";
    public static final String CMD_UNBIND = "UBND";
    public static final String CMD_XMLMSG = "XMLMSG";
    public static final int ERROR_ALLOCATE_SESSION_FAILED = 206;
    public static final int ERROR_BIND_TIMEOUT = 206;
    public static final int ERROR_FE_TRANSPORT_CLOSED = 303;
    public static final int ERROR_FIRST_PACKET_NOT_CONN = 103;
    public static final int ERROR_HOST_UNKNOWN = 100;
    public static final int ERROR_INVALID_CHID = 200;
    public static final int ERROR_INVALID_FROM = 201;
    public static final int ERROR_INVALID_XML = 205;
    public static final int ERROR_PARSE_CONN = 102;
    public static final int ERROR_PARSE_REQUEST = 204;
    public static final int ERROR_SEND_TO_GATEWAY_FAILED = 302;
    public static final int ERROR_SEND_TO_IMS_FAILED = 300;
    public static final int ERROR_SEND_TO_XMQ_FAILED = 301;
    public static final int ERROR_UNEXPECTED_REQUEST = 202;
    public static final int ERROR_UNSUPPORTED_VERSION = 101;
    public static final int ERROR_USR_NOT_BIND = 203;
    static final int HEADER_SIZE = 8;
    private static final byte HEAD_LEN_OFFSET = 2;
    public static final String ID_NOT_AVAILABLE = "ID_NOT_AVAILABLE";
    static final byte LENGTH_OFFSET = 4;
    public static final short MAGIC = -15618;
    static final byte MAGIC_OFFSET = 0;
    public static final int MAX_BLOB_SIZE = 32768;
    static final short PAYLOAD_BINARY = 2;
    private static final byte PAYLOAD_LEN_OFFSET = 4;
    public static final short PAYLOAD_PROTO = 2;
    public static final short PAYLOAD_THRIFT = 1;
    private static final byte PAYLOAD_TYPE_OFFSET = 0;
    public static final short PAYLOAD_XML = 3;
    public static final String SERVER_PING_ID = "1";
    public static final String SUBCMD_CONF = "CONF";
    private static final byte SUB_HEADER_OFFSET = 8;
    public static final short VERSION = 5;
    public static final short VERSION5 = 4;
    static final byte VERSION_OFFSET = 2;
    public static final String XIAOMI_SERVER = "xiaomi.com";
    private ChannelMessage.ClientHeader mHeader;
    String mPackageName;
    private byte[] mPayload;
    private short mPayloadType;
    private static String prefix = StringUtils.randomString(5) + Constants.ACCEPT_TIME_SEPARATOR_SERVER;
    private static long id = 0;
    private static final byte[] EMPTY = new byte[0];

    public Blob() {
        this.mPayloadType = (short) 2;
        this.mPayload = EMPTY;
        this.mPackageName = null;
        this.mHeader = new ChannelMessage.ClientHeader();
    }

    Blob(ChannelMessage.ClientHeader clientHeader, short s, byte[] bArr) {
        this.mPayloadType = (short) 2;
        this.mPayload = EMPTY;
        this.mPackageName = null;
        this.mHeader = clientHeader;
        this.mPayloadType = s;
        this.mPayload = bArr;
    }

    public static Blob from(Packet packet, String str) {
        int i;
        Blob blob = new Blob();
        try {
            i = Integer.parseInt(packet.getChannelId());
        } catch (Exception e) {
            MyLog.w("Blob parse chid err " + e.getMessage());
            i = 1;
        }
        blob.setChannelId(i);
        blob.setPacketID(packet.getPacketID());
        blob.setFrom(packet.getFrom());
        blob.setPackageName(packet.getPackageName());
        blob.setCmd(CMD_XMLMSG, null);
        try {
            blob.setPayload(packet.toXML().getBytes("utf8"), str);
            if (TextUtils.isEmpty(str)) {
                blob.setPayloadType((short) 3);
            } else {
                blob.setPayloadType((short) 2);
                blob.setCmd(CMD_SECMSG, null);
            }
        } catch (UnsupportedEncodingException e2) {
            MyLog.w("Blob setPayload err： " + e2.getMessage());
        }
        return blob;
    }

    static Blob from(ByteBuffer byteBuffer) throws IOException {
        try {
            ByteBuffer byteBufferSlice = byteBuffer.slice();
            short s = byteBufferSlice.getShort(0);
            short s2 = byteBufferSlice.getShort(2);
            int i = byteBufferSlice.getInt(4);
            ChannelMessage.ClientHeader clientHeader = new ChannelMessage.ClientHeader();
            clientHeader.mergeFrom(byteBufferSlice.array(), byteBufferSlice.arrayOffset() + 8, s2);
            byte[] bArr = new byte[i];
            byteBufferSlice.position(s2 + 8);
            byteBufferSlice.get(bArr, 0, i);
            return new Blob(clientHeader, s, bArr);
        } catch (Exception e) {
            MyLog.w("read Blob err :" + e.getMessage());
            throw new IOException("Malformed Input");
        }
    }

    public static String nextID() {
        String string;
        synchronized (Blob.class) {
            try {
                StringBuilder sb = new StringBuilder();
                sb.append(prefix);
                long j = id;
                id = 1 + j;
                sb.append(Long.toString(j));
                string = sb.toString();
            } catch (Throwable th) {
                throw th;
            }
        }
        return string;
    }

    public int getChannelId() {
        return this.mHeader.getChid();
    }

    public String getCmd() {
        return this.mHeader.getCmd();
    }

    public byte[] getDecryptedPayload(String str) {
        if (this.mHeader.getCipher() == 1) {
            return RC4Cryption.encrypt(RC4Cryption.generateKeyForRC4(str, getPacketID()), this.mPayload);
        }
        if (this.mHeader.getCipher() == 0) {
            return this.mPayload;
        }
        MyLog.w("unknow cipher = " + this.mHeader.getCipher());
        return this.mPayload;
    }

    public int getErrCode() {
        return this.mHeader.getErrCode();
    }

    public String getErrStr() {
        return this.mHeader.getErrStr();
    }

    public String getFullUserName() {
        if (!this.mHeader.hasUuid()) {
            return null;
        }
        return Long.toString(this.mHeader.getUuid()) + "@" + this.mHeader.getServer() + "/" + this.mHeader.getResource();
    }

    public String getPackageName() {
        return this.mPackageName;
    }

    public String getPacketID() {
        String id2 = this.mHeader.getId();
        if ("ID_NOT_AVAILABLE".equals(id2)) {
            return null;
        }
        if (!this.mHeader.hasId()) {
            id2 = nextID();
            this.mHeader.setId(id2);
        }
        return id2;
    }

    public byte[] getPayload() {
        return this.mPayload;
    }

    public short getPayloadType() {
        return this.mPayloadType;
    }

    public int getSerializedSize() {
        return this.mHeader.getSerializedSize() + 8 + this.mPayload.length;
    }

    public String getSubcmd() {
        return this.mHeader.getSubcmd();
    }

    public boolean hasErr() {
        return this.mHeader.hasErrCode();
    }

    public void setChannelId(int i) {
        this.mHeader.setChid(i);
    }

    public void setCmd(String str, String str2) {
        if (TextUtils.isEmpty(str)) {
            throw new IllegalArgumentException("command should not be empty");
        }
        this.mHeader.setCmd(str);
        this.mHeader.clearSubcmd();
        if (TextUtils.isEmpty(str2)) {
            return;
        }
        this.mHeader.setSubcmd(str2);
    }

    public void setFrom(long j, String str, String str2) {
        if (j != 0) {
            this.mHeader.setUuid(j);
        }
        if (!TextUtils.isEmpty(str)) {
            this.mHeader.setServer(str);
        }
        if (TextUtils.isEmpty(str2)) {
            return;
        }
        this.mHeader.setResource(str2);
    }

    public void setFrom(String str) {
        if (TextUtils.isEmpty(str)) {
            return;
        }
        int iIndexOf = str.indexOf("@");
        try {
            long j = Long.parseLong(str.substring(0, iIndexOf));
            int iIndexOf2 = str.indexOf("/", iIndexOf);
            String strSubstring = str.substring(iIndexOf + 1, iIndexOf2);
            String strSubstring2 = str.substring(iIndexOf2 + 1);
            this.mHeader.setUuid(j);
            this.mHeader.setServer(strSubstring);
            this.mHeader.setResource(strSubstring2);
        } catch (Exception e) {
            MyLog.w("Blob parse user err " + e.getMessage());
        }
    }

    public void setPackageName(String str) {
        this.mPackageName = str;
    }

    public void setPacketID(String str) {
        this.mHeader.setId(str);
    }

    public void setPayload(byte[] bArr, String str) {
        if (TextUtils.isEmpty(str)) {
            this.mHeader.setCipher(0);
            this.mPayload = bArr;
        } else {
            this.mHeader.setCipher(1);
            this.mPayload = RC4Cryption.encrypt(RC4Cryption.generateKeyForRC4(str, getPacketID()), bArr);
        }
    }

    public void setPayloadType(short s) {
        this.mPayloadType = s;
    }

    ByteBuffer toByteArray(ByteBuffer byteBuffer) {
        ByteBuffer byteBufferAllocate = byteBuffer;
        if (byteBuffer == null) {
            byteBufferAllocate = ByteBuffer.allocate(getSerializedSize());
        }
        byteBufferAllocate.putShort(this.mPayloadType);
        byteBufferAllocate.putShort((short) this.mHeader.getCachedSize());
        byteBufferAllocate.putInt(this.mPayload.length);
        int iPosition = byteBufferAllocate.position();
        this.mHeader.toByteArray(byteBufferAllocate.array(), byteBufferAllocate.arrayOffset() + iPosition, this.mHeader.getCachedSize());
        byteBufferAllocate.position(this.mHeader.getCachedSize() + iPosition);
        byteBufferAllocate.put(this.mPayload);
        return byteBufferAllocate;
    }

    public String toString() {
        return "Blob [chid=" + getChannelId() + "; Id=" + getPacketID() + "; cmd=" + getCmd() + "; type=" + ((int) getPayloadType()) + "; from=" + getFullUserName() + " ]";
    }
}

File: legacy-runtime/src/main/java/com/xiaomi/slim/Blob.kt
package com.xiaomi.slim

import android.text.TextUtils
import com.xiaomi.channel.commonutils.logger.MyLog
import com.xiaomi.mipush.sdk.Constants
import com.xiaomi.push.protobuf.ChannelMessage
import com.xiaomi.push.service.RC4Cryption
import com.xiaomi.smack.packet.Packet
import com.xiaomi.smack.util.StringUtils
import java.io.IOException
import java.io.UnsupportedEncodingException
import java.nio.ByteBuffer

open class Blob {
    private var mPayloadType: Short = PAYLOAD_BINARY.toShort()
    private var mPayload: ByteArray = EMPTY
    var mPackageName: String? = null
    private var mHeader: ChannelMessage.ClientHeader = ChannelMessage.ClientHeader()

    constructor()

    internal constructor(
        clientHeader: ChannelMessage.ClientHeader,
        payloadType: Short,
        payload: ByteArray,
    ) {
        mHeader = clientHeader
        mPayloadType = payloadType
        mPayload = payload
    }

    val channelId: Int
        get() = mHeader.chid

    val cmd: String
        get() = mHeader.cmd

    val errCode: Int
        get() = mHeader.errCode

    val errStr: String
        get() = mHeader.errStr

    val subcmd: String
        get() = mHeader.subcmd

    val payloadType: Short
        get() = mPayloadType

    open val payload: ByteArray
        get() = mPayload

    /** Backwards-compatible alias for [from]. */
    val fullUserName: String?
        get() = from

    /** Backwards-compatible accessor for [mPackageName]. */
    val packageName: String?
        get() = mPackageName

    var from: String?
        get() {
            if (!mHeader.hasUuid()) return null
            return "${mHeader.uuid}@${mHeader.server}/${mHeader.resource}"
        }
        set(value) {
            if (value.isNullOrEmpty()) return
            val iIndexOf = value.indexOf("@")
            try {
                val uuid = value.substring(0, iIndexOf).toLong()
                val iIndexOf2 = value.indexOf("/", iIndexOf)
                val server = value.substring(iIndexOf + 1, iIndexOf2)
                val resource = value.substring(iIndexOf2 + 1)
                mHeader.uuid = uuid
                mHeader.server = server
                mHeader.resource = resource
            } catch (e: Exception) {
                MyLog.w("Blob parse user err ${e.message}")
            }
        }

    var packetID: String?
        get() {
            val id = mHeader.id
            if (ID_NOT_AVAILABLE == id) return null
            if (!mHeader.hasId()) {
                val nextId = nextID()
                mHeader.id = nextId
                return nextId
            }
            return id
        }
        set(value) {
            mHeader.id = value
        }

    open val serializedSize: Int
        get() = mHeader.serializedSize + HEADER_SIZE + mPayload.size

    fun hasErr(): Boolean = mHeader.hasErrCode()

    fun getDecryptedPayload(security: String?): ByteArray {
        return when (mHeader.cipher) {
            CIPHER_RC4 -> RC4Cryption.encrypt(
                RC4Cryption.generateKeyForRC4(security!!, packetID!!),
                mPayload,
            )
            CIPHER_NONE -> mPayload
            else -> {
                MyLog.w("unknow cipher = ${mHeader.cipher}")
                mPayload
            }
        }
    }

    fun setFrom(
        uuid: Long,
        server: String,
        resource: String?,
    ) {
        if (uuid != 0L) {
            mHeader.uuid = uuid
        }
        if (!TextUtils.isEmpty(server)) {
            mHeader.server = server
        }
        if (resource.isNullOrEmpty()) return
        mHeader.resource = resource
    }

    fun setCmd(
        cmd: String,
        subcmd: String?,
    ) {
        if (TextUtils.isEmpty(cmd)) {
            throw IllegalArgumentException("command should not be empty")
        }
        mHeader.cmd = cmd
        mHeader.clearSubcmd()
        if (subcmd.isNullOrEmpty()) return
        mHeader.subcmd = subcmd
    }

    fun setPayload(
        payload: ByteArray,
        security: String?,
    ) {
        if (security.isNullOrEmpty()) {
            mHeader.cipher = CIPHER_NONE
            mPayload = payload
        } else {
            mHeader.cipher = CIPHER_RC4
            mPayload = RC4Cryption.encrypt(
                RC4Cryption.generateKeyForRC4(security, packetID!!),
                payload,
            )
        }
    }

    fun setPayloadType(type: Short) {
        mPayloadType = type
    }

    fun setChannelId(channelId: Int) {
        mHeader.chid = channelId
    }

    fun setPackageName(packageName: String?) {
        mPackageName = packageName
    }

    open fun toByteArray(byteBuffer: ByteBuffer? = null): ByteBuffer {
        val buffer = byteBuffer ?: ByteBuffer.allocate(serializedSize)
        buffer.putShort(mPayloadType)
        buffer.putShort(mHeader.cachedSize.toShort())
        buffer.putInt(mPayload.size)
        val position = buffer.position()
        mHeader.toByteArray(buffer.array(), buffer.arrayOffset() + position, mHeader.cachedSize)
        buffer.position(mHeader.cachedSize + position)
        buffer.put(mPayload)
        return buffer
    }

    override fun toString(): String {
        return "Blob [chid=$channelId; Id=$packetID; cmd=$cmd; type=${payloadType.toInt()}; from=$from ]"
    }

    companion object {
        const val MAGIC: Short = -15618
        const val VERSION: Short = 5
        const val VERSION5: Short = 4
        const val MAX_BLOB_SIZE = 32768
        const val HEADER_SIZE = 8
        const val CHCKSUM_SIZE = 4
        const val CIPHER_NONE = 0
        const val CIPHER_RC4 = 1
        const val CIPHER_AES = 2
        const val PAYLOAD_THRIFT: Short = 1
        const val PAYLOAD_BINARY: Short = 2
        const val PAYLOAD_XML: Short = 3
        const val PAYLOAD_PROTO: Short = 2

        const val CMD_BIND = "BIND"
        const val CMD_UNBIND = "UBND"
        const val CMD_CONN = "CONN"
        const val CMD_KICK = "KICK"
        const val CMD_NOTIFY = "NOTIFY"
        const val CMD_PING = "PING"
        const val CMD_SECMSG = "SECMSG"
        const val CMD_SYNC = "SYNC"
        const val CMD_CLOSE = "CLOSE"
        const val CMD_XMLMSG = "XMLMSG"
        const val SUBCMD_CONF = "CONF"

        const val CLIENT_PING_ID = "0"
        const val SERVER_PING_ID = "1"
        const val ID_NOT_AVAILABLE = "ID_NOT_AVAILABLE"
        const val XIAOMI_SERVER = "xiaomi.com"

        const val ERROR_HOST_UNKNOWN = 100
        const val ERROR_PARSE_CONN = 102
        const val ERROR_FIRST_PACKET_NOT_CONN = 103
        const val ERROR_UNSUPPORTED_VERSION = 101
        const val ERROR_INVALID_CHID = 200
        const val ERROR_INVALID_FROM = 201
        const val ERROR_UNEXPECTED_REQUEST = 202
        const val ERROR_USR_NOT_BIND = 203
        const val ERROR_PARSE_REQUEST = 204
        const val ERROR_INVALID_XML = 205
        const val ERROR_ALLOCATE_SESSION_FAILED = 206
        const val ERROR_BIND_TIMEOUT = 206
        const val ERROR_SEND_TO_IMS_FAILED = 300
        const val ERROR_SEND_TO_XMQ_FAILED = 301
        const val ERROR_SEND_TO_GATEWAY_FAILED = 302
        const val ERROR_FE_TRANSPORT_CLOSED = 303

        private const val MAGIC_OFFSET: Byte = 0
        private const val VERSION_OFFSET: Byte = 2
        private const val HEAD_LEN_OFFSET: Byte = 2
        private const val LENGTH_OFFSET: Byte = 4
        private const val PAYLOAD_TYPE_OFFSET: Byte = 0
        private const val PAYLOAD_LEN_OFFSET: Byte = 4
        private const val SUB_HEADER_OFFSET: Byte = 8

        private val EMPTY = ByteArray(0)

        private var prefix: String = StringUtils.randomString(5) + Constants.ACCEPT_TIME_SEPARATOR_SERVER
        private var id: Long = 0

        @JvmStatic
        fun nextID(): String {
            synchronized(Blob::class.java) {
                val sb = StringBuilder()
                sb.append(prefix)
                val j = id
                id = 1 + j
                sb.append(j.toString())
                return sb.toString()
            }
        }

        @JvmStatic
        fun from(
            packet: Packet,
            security: String?,
        ): Blob {
            val blob = Blob()
            val chId = try {
                packet.channelId.toInt()
            } catch (e: Exception) {
                MyLog.w("Blob parse chid err ${e.message}")
                1
            }
            blob.setChannelId(chId)
            blob.packetID = packet.packetID
            blob.from = packet.from
            blob.mPackageName = packet.packageName
            blob.setCmd(CMD_XMLMSG, null)
            try {
                blob.setPayload(packet.toXML().toByteArray(charset("utf8")), security)
                blob.mPayloadType = if (security.isNullOrEmpty()) {
                    PAYLOAD_XML
                } else {
                    blob.setCmd(CMD_SECMSG, null)
                    PAYLOAD_BINARY
                }
            } catch (e: UnsupportedEncodingException) {
                MyLog.w("Blob setPayload err: ${e.message}")
            }
            return blob
        }

        @JvmStatic
        @Throws(IOException::class)
        fun from(byteBuffer: ByteBuffer): Blob {
            return try {
                val byteBufferSlice = byteBuffer.slice()
                val s = byteBufferSlice.getShort(0)
                val s2 = byteBufferSlice.getShort(2)
                val i = byteBufferSlice.getInt(4)
                val clientHeader = ChannelMessage.ClientHeader()
                clientHeader.mergeFrom(
                    byteBufferSlice.array(),
                    byteBufferSlice.arrayOffset() + 8,
                    s2.toInt(),
                )
                val bArr = ByteArray(i)
                byteBufferSlice.position((s2 + 8).toInt())
                byteBufferSlice.get(bArr, 0, i)
                Blob(clientHeader, s, bArr)
            } catch (e: Exception) {
                MyLog.w("read Blob err :${e.message}")
                throw IOException("Malformed Input")
            }
        }
    }
}
